src/handler_base.js

```diff
@@ -73,6 +73,26 @@ export default class HandlerBase extends EventEmitter {
         this.srcSocket.on('error', this.onSrcSocketError);
     }
 
+    detach() {
```

Comment:
Can we please add a comment to explain what this function does and what is the difference from `close` ? And do we even need to add this function? Maybe we could just add the `off()` calls to `close` and skip this function altogether?

Reply:
`close` calls `socket.destroy()` and we don't want that. Or do we? If we do then we won't support `keep-alive` but only `close`.

Reply:
But if we just detach the listeners here, the pipe between `srcSocket` and `trgSocket` will remain open, and the data will keep flowing, no? That's something we don't want as reusing such socket for another connection will cause other issues. In theory we could keep the `srcSocket` open, shut down the pipe, and close the `trgSocket`, or even reuse the pipe, but I'm not sure it's worth the effort for HTTP connections only, as they comprise only a small fraction of traffic.

Maybe shutting down the sockets (after the data from target to source is surely sent) is the easiest option, even though `keep-alive` would be ignored.

Reply:
> the pipe between srcSocket and trgSocket will remain open, and the data will keep flowing, no?

In case of `handler_forward` there is no such pipe. The pipe is between `srcRequest` and `trgRequest` + `response` (trgResponse) and `srcResponse`. `detach()` is called only on resposne 'end' event. Therefore the pipe will be closed and no data will be emitted.

> In theory we could keep the srcSocket open, shut down the pipe, and close the trgSocket, or even reuse the pipe, but I'm not sure it's worth the effort for HTTP connections only, as they comprise only a small fraction of traffic.

To be honest I feel like `handler_base` is a bit too abstract. It tries to accept both tunnels and usual requests, which have different APIs under the hood. `proxy-chain` was created 3 yrs ago, so that could be best solution at that time.

> Maybe shutting down the sockets (after the data from target to source is surely sent) is the easiest option, even though keep-alive would be ignored.

I agree. Let me fix this quick.

Reply:
> Maybe shutting down the sockets (after the data from target to source is surely sent) is the easiest option, even though keep-alive would be ignored.

Actually this still doesn't fix the issue because there is no option to disable pipelining in Node.js HTTP server.

Reply:
I can move `detach` to `handler_forward` as it's used only there. Wdyt?

Reply:
thanks @szmarczak this sounds reasonable

Reply:
Moved the `detach` function to `handler_forward.js` :) Feel free to merge now.